File: agent/self_state.py
"""Self-state kernel — identity, goals, attention, beliefs, integrity.

Adapted from TheBabelDragon/self-state-kernel so throne-room can talk
to SELF without forking that repo. Canonical kernel stays there.
This copy is the adapter used inside the agent loop.
"""

from __future__ import annotations

import copy
from typing import Any, Callable

from agent.hashutil import canonical, fnv1a

CommandFn = Callable[..., Any]


class SelfStateKernel:
    def __init__(self) -> None:
        self.values: dict[str, Any] = {}
        self.commands: dict[str, CommandFn] = {}
        self.trace: list[dict[str, Any]] = []
        self.sequence = 0

        self.SET("kernel.version", "1112")
        self.SET("kernel.status", "READY")
        self.SET("identity.name", "SELF")
        self.SET("identity.role", "field-agent")
        self.SET("identity.continuity", "persistent-across-ticks")
        self.SET("goals", [
            {
                "id": "g0",
                "text": "Maintain a coherent model of the field and answer the human operator.",
                "priority": 1,
            }
        ])
        self.SET("attention.target", "chat")
        self.SET("attention.tick", 0)
        self.SET("beliefs.world", "MetaField")
        self.SET("beliefs.loop", "observe→propose→validate→commit")
        self.SET("working", {})
        self.SET("drives.coherence", 0.7)
        self.SET("drives.curiosity", 0.45)
        self.register("CMD4", self.cmd4)
        self.register("SNAPSHOT", self.snapshot)

    def register(self, name: str, fn: CommandFn) -> None:
        self.commands[name.upper()] = fn

    def peek(self, path: str, fallback: Any = None) -> Any:
        return _get_path(self.values, path, fallback)

    def snapshot(self) -> dict[str, Any]:
        return copy.deepcopy(self.values)

    def integrity(self) -> str:
        return fnv1a(canonical(self.values))

    def GET(self, path: str, fallback: Any = None) -> Any:
        value = copy.deepcopy(_get_path(self.values, path, fallback))
        self._record("GET", path, {"default": fallback}, value)
        return value

    def SET(self, path: str, value: Any) -> Any:
        _set_path(self.values, path, copy.deepcopy(value))
        self._record("SET", path, {"value": value}, value)
        return value

    def QUERY(self, query: str) -> Any:
        upper = query.upper()
        if upper in self.commands:
            result = {"type": "command", "name": upper, "registered": True}
            self._record("QUERY", query, {}, result)
            return result
        return self.GET(query)

    def RUN(self, command: str, *args: Any) -> dict[str, Any]:
        name = command.upper()
        fn = self.commands.get(name)
        if fn is None:
            error = f"Unknown command: {name}"
            self._record("RUN_ERROR", name, {"args": args}, error)
            return {"ok": False, "operation": "RUN", "error": error}
        try:
            value = fn(*args)
            rec = self._record("RUN", name, {"args": args}, value)
            return {"ok": True, "operation": name, "value": value, "trace_sequence": rec["sequence"]}
        except Exception as err:  # noqa: BLE001 — kernel must never raise through RUN
            error = str(err)
            self._record("RUN_ERROR", name, {"args": args}, error)
            return {"ok": False, "operation": name, "error": error}

    def IF(self, condition: Any, then_command: str, else_command: str | None = None) -> dict[str, Any]:
        if not isinstance(condition, bool):
            error = f"IF condition must resolve to bool; received {type(condition).__name__}"
            self._record("IF_ERROR", then_command, {"condition": str(condition)}, error)
            return {"ok": False, "operation": "IF", "error": error}
        if condition:
            return self.RUN(then_command)
        if else_command:
            return self.RUN(else_command)
        self._record("IF_FALSE", then_command, {}, False)
        return {"ok": True, "operation": "IF", "value": False}

    def cmd4(self) -> dict[str, Any]:
        return {
            "command": "CMD4",
            "action": "NORMAL_STATE",
            "special_case": False,
            "attention": self.values.get("attention"),
            "goals": self.values.get("goals"),
            "integrity": self.integrity(),
        }

    def recent_trace(self, limit: int = 12) -> list[dict[str, Any]]:
        return self.trace[-limit:]

    def _record(self, operation: str, target: str, inputs: Any, output: Any) -> dict[str, Any]:
        self.sequence += 1
        rec = {
            "sequence": self.sequence,
            "operation": operation,
            "target": target,
            "inputs": inputs,
            "output": _safe(output),
            "state_hash": self.integrity(),
        }
        self.trace.append(rec)
        if len(self.trace) > 200:
            self.trace.pop(0)
        return rec


def _get_path(obj: dict[str, Any], path: str, fallback: Any = None) -> Any:
    cur: Any = obj
    for part in path.split("."):
        if not isinstance(cur, dict):
            return fallback
        if part not in cur:
            return fallback
        cur = cur[part]
    return cur


def _set_path(obj: dict[str, Any], path: str, value: Any) -> None:
    parts = path.split(".")
    if not parts or any(not p for p in parts):
        raise ValueError(f"Invalid state path: {path}")
    cur = obj
    for part in parts[:-1]:
        nxt = cur.get(part)
        if not isinstance(nxt, dict):
            cur[part] = {}
        cur = cur[part]
    cur[parts[-1]] = value


def _safe(value: Any) -> Any:
    try:
        canonical(value)
        return value
    except TypeError:
        return str(value)


def create_self_state() -> SelfStateKernel:
    return SelfStateKernel()

File: web/src/self-state.ts
import { canonical, fnv1a } from "./hash";

export type KernelResult = {
  ok: boolean;
  operation: string;
  value?: unknown;
  error?: string;
  trace_sequence?: number;
};

export type TraceRecord = {
  sequence: number;
  operation: string;
  target: string;
  inputs: unknown;
  output: unknown;
  state_hash: string;
};

type CommandFn = (...args: unknown[]) => unknown;

function getPath(obj: Record<string, unknown>, path: string, fallback: unknown = undefined): unknown {
  let cur: unknown = obj;
  for (const part of path.split(".")) {
    if (!cur || typeof cur !== "object" || Array.isArray(cur)) return fallback;
    cur = (cur as Record<string, unknown>)[part];
    if (cur === undefined) return fallback;
  }
  return cur;
}

function setPath(obj: Record<string, unknown>, path: string, value: unknown) {
  const parts = path.split(".");
  if (!parts.length || parts.some((p) => !p)) throw new Error(`Invalid state path: ${path}`);
  let cur = obj;
  for (let i = 0; i < parts.length - 1; i++) {
    const part = parts[i]!;
    const next = cur[part];
    if (!next || typeof next !== "object" || Array.isArray(next)) {
      cur[part] = {};
    }
    cur = cur[part] as Record<string, unknown>;
  }
  cur[parts[parts.length - 1]!] = value;
}

export class SelfStateKernel {
  values: Record<string, unknown> = {};
  commands = new Map<string, CommandFn>();
  trace: TraceRecord[] = [];
  sequence = 0;

  constructor() {
    this.SET("kernel.version", "1112");
    this.SET("kernel.status", "READY");
    this.SET("identity.name", "SELF");
    this.SET("identity.role", "field-agent");
    this.SET("identity.continuity", "persistent-across-ticks");
    this.SET("goals", [
      {
        id: "g0",
        text: "Maintain a coherent model of the field and answer the human operator.",
        priority: 1,
      },
    ]);
    this.SET("attention.target", "chat");
    this.SET("attention.tick", 0);
    this.SET("beliefs.world", "MetaField");
    this.SET("beliefs.loop", "observe→propose→validate→commit");
    this.SET("working", {});
    this.SET("drives.coherence", 0.7);
    this.SET("drives.curiosity", 0.45);

    this.register("CMD4", () => this.cmd4());
    this.register("SNAPSHOT", () => this.snapshot());
  }

  register(name: string, fn: CommandFn) {
    this.commands.set(name.toUpperCase(), fn);
  }

  peek(path: string, fallback: unknown = undefined): unknown {
    let cur: unknown = this.values;
    for (const part of path.split(".")) {
      if (!cur || typeof cur !== "object" || Array.isArray(cur)) return fallback;
      cur = (cur as Record<string, unknown>)[part];
      if (cur === undefined) return fallback;
    }
    return cur;
  }

  snapshot(): Record<string, unknown> {
    return structuredClone(this.values);
  }

  integrity(): string {
    return fnv1a(canonical(this.values));
  }

  GET(path: string, fallback: unknown = undefined): unknown {
    const value = structuredClone(getPath(this.values, path, fallback));
    this.record("GET", path, { default: fallback }, value);
    return value;
  }

  SET(path: string, value: unknown): unknown {
    setPath(this.values, path, structuredClone(value));
    this.record("SET", path, { value }, value);
    return value;
  }

  QUERY(query: string): unknown {
    const upper = query.toUpperCase();
    if (this.commands.has(upper)) {
      const result = { type: "command", name: upper, registered: true };
      this.record("QUERY", query, {}, result);
      return result;
    }
    return this.GET(query);
  }

  RUN(command: string, ...args: unknown[]): KernelResult {
    const name = command.toUpperCase();
    const fn = this.commands.get(name);
    if (!fn) {
      const error = `Unknown command: ${name}`;
      this.record("RUN_ERROR", name, { args }, error);
      return { ok: false, operation: "RUN", error };
    }
    try {
      const value = fn(...args);
      const rec = this.record("RUN", name, { args }, value);
      return { ok: true, operation: name, value, trace_sequence: rec.sequence };
    } catch (err) {
      const error = err instanceof Error ? err.message : String(err);
      this.record("RUN_ERROR", name, { args }, error);
      return { ok: false, operation: name, error };
    }
  }

  IF(condition: unknown, thenCommand: string, elseCommand?: string): KernelResult {
    if (typeof condition !== "boolean") {
      const error = `IF condition must resolve to bool; received ${typeof condition}`;
      this.record("IF_ERROR", thenCommand, { condition: String(condition) }, error);
      return { ok: false, operation: "IF", error };
    }
    if (condition) return this.RUN(thenCommand);
    if (elseCommand) return this.RUN(elseCommand);
    this.record("IF_FALSE", thenCommand, {}, false);
    return { ok: true, operation: "IF", value: false };
  }

  cmd4() {
    const goals = this.values.goals;
    const attention = this.values.attention;
    return {
      command: "CMD4",
      action: "NORMAL_STATE",
      special_case: false,
      attention,
      goals,
      integrity: this.integrity(),
    };
  }

  recentTrace(limit = 12): TraceRecord[] {
    return this.trace.slice(-limit);
  }

  private record(operation: string, target: string, inputs: unknown, output: unknown): TraceRecord {
    this.sequence += 1;
    const rec: TraceRecord = {
      sequence: this.sequence,
      operation,
      target,
      inputs,
      output: safe(output),
      state_hash: this.integrity(),
    };
    this.trace.push(rec);
    if (this.trace.length > 200) this.trace.shift();
    return rec;
  }
}

function safe(value: unknown): unknown {
  try {
    JSON.stringify(value);
    return value;
  } catch {
    return String(value);
  }
}

export function createSelfState(): SelfStateKernel {
  return new SelfStateKernel();
}
